memcache/selector.go

```diff
@@ -111,19 +153,142 @@ var keyBufPool = sync.Pool{
 	},
 }
 
-func (ss *ServerList) PickServer(key string) (net.Addr, error) {
-	ss.mu.RLock()
-	defer ss.mu.RUnlock()
-	if len(ss.addrs) == 0 {
+func (ss *ServerList) PickServer(key string) (picked net.Addr, err error) {
+	ss.mu.Lock()
+	defer ss.mu.Unlock()
+
+	// if we pick a connection for retry, remove it from the available pool till we
+	// hear the result
+	defer (func() {
+		if err != nil {
+			return
+		}
+		// remove the address from available roster again to avoid multiple retries
+		if ws, ok := ss.states[picked]; ok && ws.retry == retryReady {
+			ws.retry = retryRunning
+			ss.setState(picked, ws)
+		}
+	})()
+
+	if len(ss.available) == 0 {
 		return nil, ErrNoServers
 	}
-	if len(ss.addrs) == 1 {
-		return ss.addrs[0], nil
+	if len(ss.available) == 1 {
+		return ss.available[0], nil
 	}
 	bufp := keyBufPool.Get().(*[]byte)
 	n := copy(*bufp, key)
 	cs := crc32.ChecksumIEEE((*bufp)[:n])
 	keyBufPool.Put(bufp)
 
-	return ss.addrs[cs%uint32(len(ss.addrs))], nil
+	return ss.available[cs%uint32(len(ss.available))], nil
+}
+
+func (ss *ServerList) OnResult(addr net.Addr, err error) {
+	ss.mu.Lock()
+	defer ss.mu.Unlock()
+
+	if err == nil || !isConnectionError(err) {
+		// server is considered available once we successfully
+		// communicate
+		if _, ok := ss.states[addr]; ok {
+			ss.deleteState(addr)
+		}
+		return
+	}
+
+	ws := waitState{
+		addr:  addr,
+		wait:  startingWait,
+		retry: retryWait,
+	}
+	if st, ok := ss.states[addr]; ok {
+		// We've already registered an error. Since we use connections concurrently
+		// we want to avoid double backing-off (and double retrying) here.
+		switch st.retry {
+		case retryWait:
+			// waiting for retry to be scheduled
+			return
+		case retryReady:
+			// scheduled, waiting to be picked
+			return
+		case retryRunning:
+			// backoff: we're here because we hit an error during the retry
+			ws = st
+			ws.wait *= 2
+			ws.retry = retryWait
+			if ws.wait > maxRetryWait {
+				ws.wait = maxRetryWait
+			}
+		default:
+			panic(fmt.Errorf("unexpected retry state: %s", st.retry))
+		}
+	}
+	ss.setState(addr, ws)
+
+	if ws.retry == retryWait {
+		go ss.scheduleRetry(addr, ws.wait)
+	}
+}
+
+// Sets the retry state for an address, and updates the available list.
+// MUST be called from a method with a lock on the mutex, or will cause concurrent map crashes
+func (ss *ServerList) setState(addr net.Addr, ws waitState) {
+	if ss.states == nil {
+		ss.states = map[net.Addr]waitState{}
+	}
+	ss.states[addr] = ws
+	ss.filterAvailable()
+}
+
+// MUST be called from a method with a lock on the mutex, or will cause concurrent map crashes
+func (ss *ServerList) deleteState(addr net.Addr) {
+	delete(ss.states, addr)
+	ss.filterAvailable()
+}
+
+func (ss *ServerList) filterAvailable() {
```

Comment:
is it worth testing this in isolation? same question applies to some of the other state management bits.

Reply:
👍